Add packages architecture future (#410)
vtex/faststore: docs/architecture.md

# Packages

All packages follow Gatsby's [plugin naming scheme](https://www.gatsbyjs.com/docs/plugins-themes-and-starters/). 

> * **gatsby-plugin-graphql** adds Webpack and Babel configuration so we can share queries and types between Gatsby and any other GraphQL server. This plugin is essential for dynamic, client-side queries.

Reviewer: ```suggestion
* **@vtex/gatsby-plugin-graphql** adds Webpack and Babel configuration so we can share queries and types between Gatsby and any other GraphQL server. This plugin is essential for dynamic, client-side queries.
```

> * **gatsby-plugin-i18n** is an optimized react-intl capable plugin. This plugin pre-optimizes messages we can serve react-intl without an ICU parser to the browser. Also, this plugins adds all necessary providers to the React tree

Reviewer 2: ```suggestion
* **gatsby-plugin-i18n** is an optimized react-intl capable plugin. This plugin pre-optimizes messages we can serve react-intl without an ICU parser to the browser. Also, these plugins adds all necessary providers to the React tree.
```

> * **gatsby-plugin-i18n** is an optimized react-intl capable plugin. This plugin pre-optimizes messages we can serve react-intl without an ICU parser to the browser. Also, this plugins adds all necessary providers to the React tree

Reviewer: This doesn't mean that dynamic interpolations would not happen right?

I mean, if I have a component that renders something like this based on the cart number of items:

> There are 3 products in your cart.

or

> There are 1 product in your cart.

or

> There are no products in your cart.

This should be handled by an i18n library in the client

Author: No, we still have react-intl, however, we implement [this guide](https://formatjs.io/docs/guides/advanced-usage#pre-compiling-messages) to reduce react-intl sizes and blocking. time

Reviewer: Awesome!

> * **gatsby-plugin-i18n** is an optimized react-intl capable plugin. This plugin pre-optimizes messages we can serve react-intl without an ICU parser to the browser. Also, this plugins adds all necessary providers to the React tree

Author: ```suggestion
* **gatsby-plugin-i18n** is an optimized react-intl capable plugin. This plugin pre-optimizes messages we can serve react-intl without an ICU parser to the browser. Also, this plugin adds all necessary providers to the React tree
```

> * **gatsby-plugin-i18n** is an optimized react-intl capable plugin. This plugin pre-optimizes messages we can serve react-intl without an ICU parser to the browser. Also, this plugin adds all necessary providers to the React tree

Reviewer: ```suggestion
* **@vtex/gatsby-plugin-i18n** is an optimized react-intl capable plugin. This plugin pre-optimizes messages we can serve react-intl without an ICU parser to the browser. Also, this plugin adds all necessary providers to the React tree
```

> * **gatsby-plugin-theme-ui** is an optimized theme-ui plugin for Gatsby. It includes many optmizations, not only on `theme-ui` but also on `emotion`. The goal is to merge this package into their community based versions so all the community benefits from such optimizations

Reviewer 2: ```suggestion
* **gatsby-plugin-theme-ui** is an optimized theme-ui plugin for Gatsby. It includes many optimizations, not only on `theme-ui` but also on `emotion`. The goal is to merge this package into their community-based versions so all the community benefits from such optimizations.
```

> * **gatsby-plugin-theme-ui** is an optimized theme-ui plugin for Gatsby. It includes many optimizations, not only on `theme-ui` but also on `emotion`. The goal is to merge this package into their community-based versions so all the community benefits from such optimizations.

Reviewer: ```suggestion
* **@vtex/gatsby-plugin-theme-ui** is an optimized theme-ui plugin for Gatsby. It includes many optimizations, not only on `theme-ui` but also on `emotion`. The goal is to merge this package into their community-based versions so all the community benefits from such optimizations.
```

> * **gatsby-plugin-nginx** allows you to serve your Gatsby site using `nginx`. This plugin creates a Netlify like experience where all dynamic pages, redirects and reverse proxies are merged automatically into a `nginx.conf` file. Just pick this file and start serving your site from anywhere

Reviewer 2: ```suggestion
* **gatsby-plugin-nginx** allows you to serve your Gatsby site using `nginx`. This plugin creates a Netlify like experience where all dynamic pages, redirects, and reverse proxies are merged automatically into an `nginx.conf` file. Just pick this file and start serving your site from anywhere.
```

> * **gatsby-plugin-nginx** allows you to serve your Gatsby site using `nginx`. This plugin creates a Netlify like experience where all dynamic pages, redirects, and reverse proxies are merged automatically into an `nginx.conf` file. Just pick this file and start serving your site from anywhere.

Reviewer: ```suggestion
* **@vtex/gatsby-plugin-nginx** allows you to serve your Gatsby site using `nginx`. This plugin creates a Netlify like experience where all dynamic pages, redirects, and reverse proxies are merged automatically into an `nginx.conf` file. Just pick this file and start serving your site from anywhere.
```

> * **gatsby-source-vtex** connects the Gatsby data layer to VTEX's GraphQL graph

Reviewer 2: ```suggestion
* **gatsby-source-vtex** connects the Gatsby data layer to VTEX's GraphQL.
```

> * **gatsby-source-vtex** connects the Gatsby data layer to VTEX's GraphQL.

Reviewer: ```suggestion
* **@vtex/gatsby-source-vtex** connects the Gatsby data layer to VTEX's GraphQL.
```

> * **gatsby-theme-store** contains a Web SDK and is the main skeleton of an ecommerce. In it you will find the web's best practices, like above/below the fold techniques an first class Pixel integration. Note that this repo should NOT contain any non structurant component

Reviewer 2: ```suggestion
* **gatsby-theme-store** contains a Web SDK and is the main skeleton of an ecommerce. In it, you will find the web's best practices, like above/below the fold techniques and first-class Pixel integration. Note that this repo should NOT contain any nonstructural component.
```

> * **gatsby-theme-store** contains a Web SDK and is the main skeleton of an ecommerce. In it you will find the web's best practices, like above/below the fold techniques an first class Pixel integration. Note that this repo should NOT contain any non structurant component

Reviewer: Not clear what is the API here. Maybe small pieces would be easier to maintain and do breaking changes.

Author: This package should contain basically `/sdk` `/pages` `/templates` and some structuring components in `/components`, like BelowTheFold.tsx and AboveTheFold.tsx from `gatsby-theme-vtex`

The `/sdk` should contain all basic store logic, like search pagination, sales channel, auth, queries, etc.
The `/pages` and `/templates` are the webpack entrypoints
The `/components` have  components that MUST be shadowed, like Layout, BelowTheFold.tsx and AboveTheFold.tsx

Reviewer: Does it make sense to split the sdk packages? I'm imagining a situation where we want to do a breaking change of a specific SDK, the user wouldn't need to update the basic plugin.

Also **gatsby-theme-vtex-store** isn't a better name?

Author: I'm really not sure about this one. The tradeoff I see in here is that splitting the packages give you greater flexibility, however maintaining a monolithic structure is easier for new people joining and makes us keep the codebase small. However, I really don't know.

About the package naming scheme, I think these packages can be the beginning of the OpenStore initiative

> * **gatsby-theme-store** contains a Web SDK and is the main skeleton of an ecommerce. In it, you will find the web's best practices, like above/below the fold techniques and first-class Pixel integration. Note that this repo should NOT contain any nonstructural component.

Reviewer: ```suggestion
* **@vtex/gatsby-theme-store** contains a Web SDK and is the main skeleton of an ecommerce. In it, you will find the web's best practices, like above/below the fold techniques and first-class Pixel integration. Note that this repo should NOT contain any nonstructural component.
```

> * **gatsby-plugin-checkout** implements checkout's Web SDK. It can be used without `gatsby-theme-store` to create extremelly custom experiences

Reviewer 2: ```suggestion
* **gatsby-plugin-checkout** implements checkout's Web SDK. It can be used without `gatsby-theme-store` to create extremely custom experiences.
```

> * **gatsby-plugin-checkout** implements checkout's Web SDK. It can be used without `gatsby-theme-store` to create extremelly custom experiences

Reviewer: Is that related to VTEX's Checkout?

Author: Can be ! The idea is that this plugin will have all orderForm related APIs and maybe the checkout flow. Maybe it's better for this to be **gatsby-theme-checkout**, what do you think ?

Reviewer: Shouldn't those package have vtex in its name? gatsby-theme-vtex-checkout

Author: I don't think so, do you know about the "OpenStore" project ?

Reviewer 2: I like the suggestion of @klzns 

> gatsby-theme-vtex-checkout

Author: Also, there is something I forgot to say. All packages are already scoped by the `@vtex` namespace. So the full name would be: `@vtex/gatsby-theme-checkout`. That's why I think we don't need the `-vtex-` in the middle of the name, like in `@vtex/gatsbty-theme-vtex-checkout`

Reviewer: oh, okk, that's make sense!

> * **gatsby-plugin-checkout** implements checkout's Web SDK. It can be used without `gatsby-theme-store` to create extremely custom experiences.

Reviewer: ```suggestion
* **@vtex/gatsby-plugin-checkout** implements checkout's Web SDK. It can be used without `gatsby-theme-store` to create extremely custom experiences.
```

> * **gatsby-plugin-vtex-cms** integrates VTEX CMS with Gatsby. Live preview and pilot your ecommerce directly from VTEX Admin. Note that this overwrites any page in your gatsby project

Reviewer 2: ```suggestion
* **gatsby-plugin-vtex-cms** integrates VTEX CMS with Gatsby. Live preview and pilot your ecommerce directly from VTEX Admin. Note that this overwrites any page in your gatsby project.
```

> * **gatsby-plugin-vtex-cms** integrates VTEX CMS with Gatsby. Live preview and pilot your ecommerce directly from VTEX Admin. Note that this overwrites any page in your gatsby project.

Reviewer: ```suggestion
* **@vtex/gatsby-plugin-vtex-cms** integrates VTEX CMS with Gatsby. Live preview and pilot your ecommerce directly from VTEX Admin. Note that this overwrites any page in your gatsby project.
```

> * **@vtex/gatsby-plugin-vtex-cms** integrates VTEX CMS with Gatsby. Live preview and pilot your ecommerce directly from VTEX Admin. Note that this overwrites any page in your gatsby project.

Reviewer 2: Doesn't using the word `vtex` leave the plugin name redundant since it already has the **@vtex** prefix?

```suggestion
* **@vtex/gatsby-plugin-cms** integrates VTEX CMS with Gatsby. Live preview and pilot your ecommerce directly from VTEX Admin. Note that this overwrites any page in your gatsby project.
```

> * **lighthouse-config** contains an ecommerce focused Lighthouse CI configuration

Reviewer 2: ```suggestion
* **lighthouse-config** contains an ecommerce focused Lighthouse CI configuration.
```

> * **lighthouse-config** contains an ecommerce focused Lighthouse CI configuration.

Reviewer: ```suggestion
* **@vtex/lighthouse-config** contains an ecommerce focused Lighthouse CI configuration.
```

> * **store-ui** contains many building blocks for creating your store. All components follows [this pattern](/components.md). Use it's storybook to find the right component for your ecommerce experience

Reviewer 2: ```suggestion
* **store-ui** contains many building blocks for creating your store. All components follows [this pattern](/components.md). Use it's storybook to find the right component for your ecommerce experience.
```

> * **store-ui** contains many building blocks for creating your store. All components follows [this pattern](/components.md). Use it's storybook to find the right component for your ecommerce experience.

Reviewer: ```suggestion
* **@vtex/store-ui** contains many building blocks for creating your store. All components follows [this pattern](/components.md). Use it's storybook to find the right component for your ecommerce experience.
```
